### PR TITLE
Filetracer Plugin for Linux

### PR DESCRIPTION
**Pull request by Dos98**
Hi Everyone,
I have completed the filetracer plugin for Linux os. I would like you review it and give suggestions to improve it in every-way.

> I have restructured the filetracer class.
> I tried to capture almost all file operations for linux file tracer.

I have tested it for Linux Ubuntu 20.04 and Debian Buster.

### CODE REVIEW

**Comment by Dos98** [comment]
For Ubuntu 20.04, I see that gcc optimizes kernel sym utimes_common to utimes_common.isra.0 and not in other linux distros. Therefore it would fail to register trap there on other linux distros.

cat /proc/kallsyms | grep utimes_common
Just shorten the trap's syscall name.

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

File: src/plugins/filetracer/filetracer.cpp [hunk 1]
```diff
@@ -103,940 +103,29 @@
...
-    trap->data = w;
-    trap->cb = create_file_ret_cb;
-
-    if ( !drakvuf_add_trap(drakvuf, trap) )
+    this->os = drakvuf_get_os_type(drakvuf);
+    if (this->os == VMI_OS_WINDOWS)
     {
-        printf("Failed to trap return at 0x%lx\n", rsp);
-        delete w;
+        new win_filetracer(drakvuf, output);
```
Comment (tklengyel) [inline src/plugins/filetracer/filetracer.cpp:121]: Valgrind reports memory leak, this never gets `delete`-d when shutting down the plugin

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
On the debian-stretch test VM filetracer fails: `Plugin filetracer startup failed!`

**Comment by Dos98** [comment]
Commented for now, as kernel symbol's suffix varies with distros.


> For Ubuntu 20.04, I see that gcc optimizes kernel sym utimes_common to utimes_common.isra.0 and not in other linux distros. Therefore it would fail to register trap there on other linux distros.
> 
> cat /proc/kallsyms | grep utimes_common
> Just shorten the trap's syscall name.
> File: filetracer.cpp in register_trap function call

**Comment by tklengyel** [comment]
On the debian-stretch kernel I have
```
root@debian-stretch:~# cat /proc/kallsyms | grep utimes_common
ffffffff9549fb10 t utimes_common
```

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by Dos98** [comment]
@tklengyel  Let me know what was the error reported by drakvuf-jenkins.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
Memory leaks caught by valgrind:
```
 0264             dirname = g_strdup_printf("%s", vmi_read_str(vmi, &ctx));
```

```
 0315     lw->filename = g_strdup_printf("%s", get_filepath(drakvuf, info, vmi_lg.vmi, dentry_addr).data());
```

and a bunch more. Please use valgrind on your local machine and fix these.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
Looks like you have a double-free bug somewhere: `free(): invalid pointer`

File: .gitmodules [hunk 2]
```diff
@@ -1,7 +1,3 @@
-[submodule "xen"]
```
Comment (tklengyel) [inline .gitmodules:1]: Why did you remove this submodule?

**Comment by tklengyel** [comment]
Please don't change the `libvmi` submodule, if it needs to be updated that's usually done in separate PRs

File: .gitmodules (hunk 2 above)
Comment (Dos98) [reply .gitmodules]: I updated my xen version and made some changes., And forgot to review while making a update on PR..

**Comment by Dos98** [comment]
Certainly prev 6 commits looks like a git blunder. Confused by git revert. Fixed .gitignore. It's all good now. phew!

**Comment by tklengyel** [comment]
@Dos98 I tested the PR and it works great! Would need a final rebase to resolve some conflicts due to some recent changes and then I'm happy to merge. Thanks for adding this! 👍 🥇
